presto-orc/src/main/java/com/facebook/presto/orc/OrcWriteValidation.java

```diff
@@ -236,35 +255,146 @@ public void validateRowGroupStatistics(OrcDataSourceId orcDataSourceId, long str
+                // have 0 values before marking them as excluded.
+                Integer mapNode = keyToMapNodeEntry.getValue();
+                Set<Integer> valueNodes = flattenedMapToValueNodes.get(mapNode);
+                for (int valueNode : valueNodes) {
+                    ColumnStatistics valueColumnStat = allStripeColumnStatistics.get(valueNode);
+                    if (valueColumnStat.getNumberOfValues() != 0) {
+                        throw new OrcCorruptionException(
+                                orcDataSourceId,
+                                "Stripe at offset %s has unexpected flat map value node column stats with non-zero number of values",
+                                stripeOffset);
+                    }
+                    excludedNodes.add(valueNode);
+                }
+            }
+        }
+
+        // return original stats excluding the flat map value stats for empty flat maps
+        return stats.entrySet().stream()
+                .filter(e -> !excludedNodes.contains(e.getKey()))
+                .collect(toImmutableMap(Entry::getKey, Entry::getValue));
+    }
+
+    private Map<Integer, ColumnStatistics> aggregateRowGroupStatisticsFromRowIndex(
+            OrcDataSourceId orcDataSourceId,
+            Map<StreamId, List<RowGroupIndex>> actualRowGroupStatistics,
+            long stripeOffset,
+            int rowGroupIndex)
+    {
+        // flattened nodes might have multiple ROW_INDEX with the same column, but different sequences
+        // aggregate such statistics before the validation
+        Map<Integer, List<ColumnStatistics>> actualColumnStatisticsByColumn = new HashMap<>();
+        for (Entry<StreamId, List<RowGroupIndex>> entry : actualRowGroupStatistics.entrySet()) {
+            int column = entry.getKey().getColumn();
+            ColumnStatistics actual = entry.getValue().get(rowGroupIndex).getColumnStatistics();
+            List<ColumnStatistics> aggregateStats = actualColumnStatisticsByColumn.computeIfAbsent(column, (key) -> new ArrayList<>());
+            aggregateStats.add(actual);
+
+            // Regular nodes have only 1 ColumnStatistics in the ROW_INDEX, flattened nodes
+            // might have zero or more column statistics.
+            if (aggregateStats.size() != 1 && !flattenedValueNodes.contains(column)) {
+                throw new OrcCorruptionException(
+                        orcDataSourceId,
+                        "Unexpected multiple column statistics for node %s in row group %s in stripe at offset %s",
+                        column,
+                        rowGroupIndex,
+                        stripeOffset);
+            }
+        }
+
+        return actualColumnStatisticsByColumn.entrySet().stream()
+                .collect(Collectors.toMap(Entry::getKey, entry -> mergeColumnStatistics(entry.getValue())));
+    }
+
+    private Map<Integer, Integer> getFlattenedKeyToMapNodes(Set<Integer> flattenedNodes, List<OrcType> orcTypes)
```

Comment:
isn't this always +1 to the node id ? As flat Maps only support primitive types.

Reply:
Why make assumptions instead of following how the tree is actually built? `Yes` for OrcType created from Presto types in the writer constructor, may be `no` for OrcType created by outside of the writer constructor.